File: pegasusX/apps/warehouse-app-ios/WarehouseApp/Views/Replenishment/ReplenishmentList.swift
import SwiftUI

struct ReplenishmentList: View {
    let insights: [ReplenishmentInsight]
    let actingId: String?
    let onApprove: (String) -> Void
    let onDismiss: (String) -> Void

    var body: some View {
        ResponsiveGridContentWrapper {
            ForEach(insights) { insight in
                VStack(alignment: .leading, spacing: LabTheme.spacingSM) {
                    HStack(spacing: LabTheme.spacingXS) {
                        Text(insight.productName)
                            .font(.headline)
                        if insight.reasonCode == "PREDICTIVE_PUSH" {
                            Text("mobile_warehouse.ui.ai_push")
                                .font(.system(size: 10, weight: .bold))
                                .padding(.horizontal, 6)
                                .padding(.vertical, 2)
                                .background(Color.primary)
                                .foregroundColor(Color(UIColor.systemBackground))
                                .clipShape(Capsule())
                        }
                    }
                    HStack(spacing: LabTheme.spacingSM) {
                        WarehouseStatusBadge(text: insight.urgency, tint: urgencyTint(insight.urgency))
                        WarehouseStatusBadge(text: insight.status)
                    }
                    Text(L10n.format("mobile_warehouse.ui.stock_currentstock_reorder_reorderquantity_2", "\(insight.currentStock)", "\(insight.reorderQuantity)"))
                        .font(.caption)
                        .foregroundStyle(.secondary)
                    Text(L10n.format("mobile_warehouse.ui.days_until_stockout_daysuntilstockout", "\(insight.daysUntilStockout)"))
                        .font(.caption)
                        .foregroundStyle(.secondary)
                    if let why = demandWhyText(insight) {
                        Text(why)
                            .font(.caption)
                            .foregroundStyle(.secondary)
                    }
                    if insight.status.uppercased() == "OPEN" {
                        HStack {
                            Button("mobile_warehouse.ui.approve") {
                                onApprove(insight.id)
                            }
                            .disabled(actingId == insight.id)
                            Button("factory_portal.toast.text.dismiss", role: .destructive) {
                                onDismiss(insight.id)
                            }
                            .disabled(actingId == insight.id)
                        }
                        .buttonStyle(.bordered)
                    }
                }
                .padding(.vertical, LabTheme.spacingXS)
            }
        }
    }

    private func urgencyTint(_ urgency: String) -> Color {
        switch urgency.uppercased() {
        case "PROACTIVE": return .primary
        case "CRITICAL": return LabTheme.destructive
        case "URGENT": return LabTheme.warning
        default: return LabTheme.secondaryLabel
        }
    }

    private func demandWhyText(_ insight: ReplenishmentInsight) -> String? {
        if let breakdown = insight.demandBreakdown, !breakdown.isEmpty {
            if let blocked = string(from: breakdown["blocked_reason"]), !blocked.isEmpty {
                return blocked == "insufficient_history"
                    ? "Insufficient history — forecast blocked"
                    : blocked.replacingOccurrences(of: "_", with: " ")
            }
            var parts: [String] = []
            if let burn = number(from: breakdown["burn_rate_7d"]) ?? number(from: breakdown["burn_rate"]) {
                parts.append(String(format: "Burn %.1f/d", burn))
            }
            if let cover = number(from: breakdown["days_cover"]) {
                parts.append(String(format: "%.1fd cover", cover))
            }
            if breakdown["mei_network"] != nil {
                parts.append("MEIO network transfer")
            }
            if !parts.isEmpty { return parts.joined(separator: " · ") }
        }
        if let code = insight.reasonCode, !code.isEmpty {
            return code.replacingOccurrences(of: "_", with: " ")
        }
        return nil
    }

    private func number(from codable: AnyCodable?) -> Double? {
        guard let codable else { return nil }
        if let d = codable.value as? Double { return d }
        if let i = codable.value as? Int { return Double(i) }
        if let s = codable.value as? String, let d = Double(s) { return d }
        return nil
    }

    private func string(from codable: AnyCodable?) -> String? {
        guard let codable else { return nil }
        if let s = codable.value as? String, !s.isEmpty { return s }
        return nil
    }
}

File: pegasusX/apps/warehouse-app-android/app/src/main/java/com/pegasusx/warehouse/ui/screens/replenishment/ReplenishmentList.kt
package com.pegasusx.warehouse.ui.screens.replenishment

import androidx.compose.ui.res.stringResource

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.GridItemSpan
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.pegasusx.warehouse.data.model.ReplenishmentInsight
import com.pegasusx.warehouse.ui.components.WarehouseSectionTitle
import com.pegasusx.warehouse.ui.components.WarehouseStatusChip
import com.pegasusx.warehouse.ui.theme.PegasusSpacing
import com.pegasusx.warehouse.R

@Composable
fun ReplenishmentList(
    insights: List<ReplenishmentInsight>,
    actingId: String?,
    onApprove: (String) -> Unit,
    onDismiss: (String) -> Unit,
    modifier: Modifier = Modifier
) {
    LazyVerticalGrid(
        columns = GridCells.Adaptive(minSize = 340.dp),
        modifier = modifier.fillMaxSize(),
        contentPadding = PaddingValues(PegasusSpacing.lg),
        verticalArrangement = Arrangement.spacedBy(PegasusSpacing.md),
        horizontalArrangement = Arrangement.spacedBy(PegasusSpacing.md),
    ) {
        item(span = { GridItemSpan(maxLineSpan) }) {
            WarehouseSectionTitle("Open insights (${insights.size})")
        }
        items(insights, key = { it.id }) { insight ->
            InsightCard(
                insight = insight,
                busy = actingId == insight.id,
                onApprove = { onApprove(insight.id) },
                onDismiss = { onDismiss(insight.id) },
            )
        }
    }
}

@Composable
internal fun InsightCard(
    insight: ReplenishmentInsight,
    busy: Boolean,
    onApprove: () -> Unit,
    onDismiss: () -> Unit,
) {
    ElevatedCard(Modifier.fillMaxWidth()) {
        Column(Modifier.padding(PegasusSpacing.lg), verticalArrangement = Arrangement.spacedBy(PegasusSpacing.sm)) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Row(modifier = Modifier.weight(1f), verticalAlignment = Alignment.CenterVertically) {
                    Text(insight.productName, style = MaterialTheme.typography.titleMedium)
                    if (insight.reasonCode == "PREDICTIVE_PUSH") {
                        Spacer(Modifier.width(PegasusSpacing.xs))
                        Surface(
                            color = MaterialTheme.colorScheme.onSurface,
                            shape = MaterialTheme.shapes.small
                        ) {
                            Row(
                                verticalAlignment = Alignment.CenterVertically,
                                modifier = Modifier.padding(horizontal = 6.dp, vertical = 2.dp)
                            ) {
                                Text(
                                    "AI PUSH",
                                    style = MaterialTheme.typography.labelSmall.copy(fontSize = 10.sp),
                                    color = MaterialTheme.colorScheme.surface
                                )
                            }
                        }
                    }
                }
                WarehouseStatusChip(status = insight.urgency)
            }
            Row(horizontalArrangement = Arrangement.spacedBy(PegasusSpacing.sm)) {
                WarehouseStatusChip(status = insight.status)
            }
            Text(
                stringResource(R.string.mobile_warehouse_ui_stock_currentstock_reorder_reorderquantity, insight.currentStock, insight.reorderQuantity),
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
            insight.demandBreakdown?.let { breakdown ->
                Text(formatDemandWhy(breakdown, insight.reasonCode), style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
            }
            Text(
                stringResource(R.string.mobile_warehouse_ui_days_until_stockout_daysuntilstockout, insight.daysUntilStockout),
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
            if (insight.status.equals("OPEN", ignoreCase = true)) {
                Row(horizontalArrangement = Arrangement.spacedBy(PegasusSpacing.sm)) {
                    Button(onClick = onApprove, enabled = !busy) { Text("Approve") }
                    OutlinedButton(onClick = onDismiss, enabled = !busy) { Text("Dismiss") }
                }
            }
        }
    }
}

internal fun formatDemandWhy(breakdown: kotlinx.serialization.json.JsonObject?, reasonCode: String?): String {
    if (breakdown == null || breakdown.isEmpty()) {
        return reasonCode?.replace('_', ' ') ?: "Threshold breach"
    }
    breakdown["blocked_reason"]?.toString()?.trim('"')?.takeIf { it.isNotBlank() }?.let { blocked ->
        return if (blocked == "insufficient_history") {
            "Insufficient history — forecast blocked"
        } else {
            blocked.replace('_', ' ')
        }
    }
    val parts = mutableListOf<String>()
    breakdown["burn_rate_7d"]?.toString()?.trim('"')?.toDoubleOrNull()?.let { parts.add("Burn ${"%.1f".format(it)}/d") }
    breakdown["days_cover"]?.toString()?.trim('"')?.toDoubleOrNull()?.let { parts.add("${"%.1f".format(it)}d cover") }
    if (breakdown.containsKey("mei_network")) parts.add("MEIO network transfer")
    return parts.joinToString(" · ").ifBlank { reasonCode?.replace('_', ' ') ?: "Demand signal" }
}
